README.md
[![PyPI version](https://badge.fury.io/py/rapoc.svg)](https://badge.fury.io/py/rapoc)
![GitHub release (latest by date)](https://img.shields.io/github/v/release/ExObsSim/Rapoc-public?color=gree&label=GitHub%20release)
[![Downloads](https://pepy.tech/badge/rapoc)](https://pepy.tech/project/rapoc)
[![Documentation Status](https://readthedocs.org/projects/rapoc-public/badge/?version=latest)](https://rapoc-public.readthedocs.io/en/latest/?badge=latest)
![GitHub](https://img.shields.io/github/license/ExObsSim/Rapoc-public)

# RAPOC: Rosseland And Planck Opacity Converter

The `RAPOC` code is written by Lorenzo V. Mugnai and Darius Modirrousta-Galian and is the product 
of a collaboration between Sapienza Università di Roma, Università degli Studi di Palermo and 
INAF - Osservatorio Astronomico di Palermo. It uses molecular absorption measurements 
(i.e. wavelength-dependent opacities) to calculate Rosseland and 
Planck mean opacities that are commonly used in atmospheric modelling.

`RAPOC` is designed to be simple, straightforward, and easily incorporated 
into other codes. It is completely written in Python and documented with docstrings. 
In addition, a Sphinx version of the documentation with a full user guide 
that includes examples is available in html format.

### Reports
`RAPOC` is under development, please report any issues or inaccuracies 
to the developers to support the implementation.

### Cite
If you use this code or its results, please cite `RAPOC: the Rosseland and Planck opacity converter` by Mugnai L. V. and Modirrousta-Galian D. (submitted).

## Installation
### Installing from Pypi
`RAPOC` can be installed from the Pypi repository with the following script::

    pip install rapoc

### Installing from git
`RAPOC` may also be cloned from the main git repository::

    git clone https://github.com/ExObsSim/Rapoc-public.git

The next step is to move into the `RAPOC` folder::

    cd /your_path/Rapoc

Then::

    pip install .

To check if one has the correct setup::

    python -c "import rapoc"


## Use
`RAPOC` is designed to be used on its own or in conjunction with other Python 
codes. Given an ExoMol file in the TauREx.h5 format, Rosseland and Planck mean opacities can be calculated. 
For example, in order to estimate the mean opacities at a temperature (T) of 1000 K with a pressure (P) of 
10,000 Pa in the wavelength range of  0.3-50 micron the following script is used,
    
    from rapoc import Rosseland, Planck

    r_model = Rosseland(input_data='exomol_file.TauREx.h5')
    opacity = r_model.estimate(P_input=10000 * u.Pa, T_input=1000 * u.K, band=(0.3 *u.um, 50*u.um))

    p_model = Planck(input_data='exomol_file.TauREx.h5')
    opacity = p_model.estimate(P_input=10000 * u.Pa, T_input=1000 * u.K, band=(0.3 *u.um, 50*u.um))

### Inputs
To run the code you need measured data. The supported file formats are:

- ExoMol opacities (downloadable [here](http://exomol.com/data/data-types/opacity)) with the `TauREx.h5` format.
- Dace opacities (downloadable [here](https://dace.unige.ch/opacityDatabase/)) with the binary format.

## Documentation
The full documentation is available [here](https://rapoc-public.readthedocs.io/en/latest/) 

Alternatively, `RAPOC` accepts user-defined documentation by using `sphinx`. To install it run
    
    pip install sphinx sphinx_rtd_theme nbsphinx
    
From the `Rapoc/docs` folder running
    
    cd docs
    make html

This will create a html version of the documentation in `Rapoc/doc/build/html/index.html`.
 

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres
to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.8] - 2022-10-6
### Added
- forced molecule name in Loader with *molecule* keyword
### Changed
- increased loading performance for Dace data
### Fixed
- ascii conversion for strings in ExoMol data

## [1.0.5] - 2021-05-11
### Added
- support for DACE opacities original data

## [1.0.4] - 2021-05-04
### Added
- added Rayleigh scattering converter. This class computes the Rayleigh scattering for the indicated atom and can be used as input data for the RAPOC models.
### Changed
- we removed "cubic" mode and added "loglinear" instead. This mode is the same as "linear" but uses the logarithm of pressure instead of simple pressure.
- we added ``self.exceptions`` variable to model. If it is not `None`, ``model.estimate`` and the validation procedures behave accordingly. The variable is given by the ``loader.read_content`` and is ``None`` by default.
- _ktable_ name replaced with _opacities_

## [1.0.0] - 2021-03-02
First RAPOC release


[Unreleased]: https://github.com/ExObsSim/Rapoc
[1.0.8]: https://github.com/ExObsSim/Rapoc-public/compare/v1.0.5...v1.0.8
[1.0.5]: https://github.com/ExObsSim/Rapoc-public/compare/v1.0.4...v1.0.5
[1.0.4]: https://github.com/ExObsSim/Rapoc-public/compare/v1.0.0...v1.0.4
[1.0.0]: https://github.com/ExObsSim/Rapoc-public/releases/tag/v1.0.0

rapoc/loaders/dictLoader.py
from astropy import units as u


class DictLoader:
    '''
    Dict loader class. The dictionary should contain the following information under certain keys:

    .. csv-table::
        :header: information, default units, supported keys

        molecular name, ,"mol"
        pressure grid, :math:`Pa`,"p, P, pressure, pressure_grid"
        temperature grid, :math:`K`,"t, T, temperature, temperature_grid"
        wavenumber grid, :math:`1/cm`,"wn, wavenumber, wavenumbers, wavenumbers_grid, wavenumber_grid"
        opacities, :math:`m^2/kg`, "opacities"

    If the input data has no units attached, the defaults units are assume.
    If they have units, a simple conversion is performed by the code to match the default values.


    Notes
    -----
    This class is implicitly called by the model when initialised if the matched input is used.

    Examples
    ---------
    First we prepare a data dictionary. Please be aware that this are not representative of any molecule.

    >>> import numpy as np
    >>> data_dict = {'mol': 'None',
    >>>              'mol_mass':42,
    >>>              'pressure': np.array([1.00000000e+00, 1.00000000e+01, 1.00000000e+02, 1.00000000e+03,
    >>>                                   1.00000000e+04, 1.00000000e+05, 1.00000000e+06, 1.00000000e+07]),
    >>>              'temperature': np.array([500, 1000, 1500, 2000, 3000]),
    >>>              'wavenumber': np.array([100000, 1000])}
    >>> opac = np.ones((data_dict['pressure'].size,data_dict['temperature'].size,data_dict['wavenumber'].size,))
    >>> data_dict['opacities'] = opac

    Let's build the Planck method using the dictionary as input

    >>> from rapoc import Planck
    >>> model = Planck(input_data=data_dict)

    Now the model is ready to be used
    '''

    def __init__(self, input_data):
        '''

        Parameters
        ----------
        input_data: dict
            input_data dictionary
        '''
        self.input_data = input_data

    def read_content(self):
        '''
        Reads the dict content and returns the needed valued for the opacity models.

        Returns
        -------
        str
            molecule name
        astropy.units.Quantity:
            molecular mass
        astropy.units.Quantity:
            data pressure grid in si units
        astropy.units.Quantity:
            data temperature grid in si units
        astropy.units.Quantity:
            data wavenumber grid
        astropy.units.Quantity:
            data opacities grid in si units
        '''
        global pressure_grid, temperature_grid, wavenumber_grid

        try:
            mol = self.input_data['mol']
        except KeyError:
            raise KeyError('molecule name not found')

        try:
            mol_mass = self.input_data['mol_mass']
        except KeyError:
            from molmass import Formula
            f = Formula(mol)
            mol_mass = (f.mass * u.u / u.ct).to(u.g / u.ct)

        if any(key in self.input_data.keys() for key in ['p', 'P', 'pressure', 'pressure_grid']):
            for press_key in ['p', 'P', 'pressure', 'pressure_grid']:
                try:
                    pressure_grid = self.input_data[press_key]
                except KeyError:
                    pass
            if isinstance(pressure_grid, u.Quantity):
                pressure_grid = pressure_grid.to(u.Pa)
            else:
                pressure_grid *= u.Pa
        else:
            raise KeyError('Pressure grid not found')

        if any(key in self.input_data.keys() for key in ['t', 'T', 'temperature', 'temperature_grid']):
            for temp_key in ['t', 'T', 'temperature', 'temperature_grid']:
                try:
                    temperature_grid = self.input_data[temp_key]
                except KeyError:
                    pass
            if isinstance(temperature_grid, u.Quantity):
                temperature_grid = temperature_grid.to(u.K)
            else:
                temperature_grid *= u.K
        else:
            raise KeyError('Temperature grid not found')

        if any(key in self.input_data.keys()
               for key in ['wn', 'wavenumber', 'wavenumbers', 'wavenumbers_grid', 'wavenumber_grid']):
            for wn_key in ['wn', 'wavenumber', 'wavenumbers', 'wavenumbers_grid', 'wavenumber_grid']:
                try:
                    wavenumber_grid = self.input_data[wn_key]
                except KeyError:
                    pass
            if isinstance(wavenumber_grid, u.Quantity):
                wavenumber_grid = wavenumber_grid.to(1 / u.cm)
            else:
                wavenumber_grid /= u.cm
        else:
            raise KeyError('Wavenumber grid not found')

        try:
            opacities = self.input_data['opacities']
            if isinstance(opacities, u.Quantity):
                opacities = opacities.to(u.m ** 2 / u.kg)
            else:
                opacities *= u.m ** 2 / u.kg
        except KeyError:
            raise KeyError('opacities not found')

        return mol, mol_mass, pressure_grid, temperature_grid, wavenumber_grid, opacities, False
